content/docs/pipelines/sdk/component-development.md

```diff
@@ -1,151 +1,67 @@
 
-Note that the example below accepts both a URI for uploading the data into, and
-a file path to write that URI to.
+#### Advanced: Producing data in an external system
 
-```
-program.py --out-model-uri gs://bucket/163/output_model \
-           --out-model-uri-file /outputs/output_model_uri/data
-```
-
-Why should the program output the URI it has just received as an input argument?
-The reason is that the URIs specified in the pipeline are usually not the real
-URIs, but rather URI templates containing UIDs. The system resolves the URIs at
-runtime when the containerized program starts. Only the containerized program 
-sees the fully-resolved URI.
-
-Below is an example of such a URI:
+In some scenarios, the goal of the component is to create some data object in an external service (E.g. a BigQuery table).
+In thais case the component program should do that and then output some identifier of the produced data location (e.g. BigQuery table name) instead of the data itself.
+This scenario should be limited to cases where the data must be put into external system instead of keeping it inside the Pipelines system.
+The class of components where this behavior is common is exporters (e.g. "upload data to GCS").
+Note that the Pipelines system cannot provide consistency and reproducibility guarantees for the data outside of its control.
 
-```
-gs://my-bucket/{{workflow.uid}}/{{pod.id}}/data
-```
+### Consuming data
 
-In cases where the program cannot control the URI/ID of the created object (for
-example, where the URI is generated by the outside system), the program should
-just accept the file path to write the resulting URI/ID:  
-
-```
-program.py --out-model-uri-file /outputs/output_model_uri/data
-```
-<a id="future-proof">
-### Future-proofing your code
-
-The following guidelines help you avoid the need to modify the program code in
-the near future or have different versions for different storage systems.
-
-If the program has access to the TensorFlow package, you can use
-[`tf.gfile`](https://www.tensorflow.org/api_docs/python/tf/io/gfile/GFile) 
-to read and write files. The `tf.gfile` module supports both local and Cloud
-Storage paths.
-
-If you cannot use `tf.gfile`, a solution is to read inputs from and write 
-outputs to local files, then add a storage-specific wrapper that downloads and 
-uploads the data from/to a specific storage solution such as
-[Cloud Storage](https://cloud.google.com/storage/docs/) or 
-[Amazon S3](https://aws.amazon.com/s3/).
-For example, create a wrapper script that uses the [`gsutil
-cp`](https://cloud.google.com/storage/docs/gsutil/commands/cp) command to
-download the input data before running the main program and to upload the output
-data after the program finishes.
+There are two main ways a command-line program usually consumes data:
+* Small pieces of data are usually passed as command-line arguments: `program.py --param1 100`.
```

Comment:
List rendering is broken here. To fix, add an empty line before the list.

See preview:https://deploy-preview-1277--competent-brattain-de2d6d.netlify.com/docs/pipelines/sdk/component-development/

Reply:
Fixed